jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java

```diff
@@ -486,60 +239,193 @@ else if (!Files.isDirectory(path) && !FileID.isWebArchive(path))
             }
 
             //set a backup value for the path to the war in case it hasn't already been set
-            properties.put(Deployable.WAR, path.toString());
-            return initializeContextHandler(context, path, properties);
+            appAttributes.setAttribute(Deployable.WAR, path.toString());
+            return initializeContextHandler(context, path, appAttributes);
         }
         finally
         {
             Thread.currentThread().setContextClassLoader(old);
         }
     }
 
-    protected Object applyXml(Object context, Path xml, String environment, Map<String, String> properties) throws Exception
+    /**
+     * Get the default {@link Environment} name for discovered web applications that
+     * do not declare the {@link Environment} that they belong to.
+     *
+     * <p>
+     * Falls back to {@link Environment#getAll()} list, and returns
+     * the first name returned after sorting with {@link Deployable#ENVIRONMENT_COMPARATOR}
+     * </p>
+     *
+     * @return the default environment name.
+     */
+    public String getDefaultEnvironmentName()
+    {
+        if (defaultEnvironmentName == null)
+        {
+            return Environment.getAll().stream()
```

Comment:
Can this be done statically once?

Reply:
The Environment list isn't static after startup, so this was the best choice for that reality.
A project could easily start with only ee10 support, then add ee8 support at runtime.
Since this is only applied at deployment, this kind of optimization wasn't deemed important.